File: app/src/main/java/com/boardgamegeek/service/SyncCollectionModifiedSince.java
package com.boardgamegeek.service;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.content.Context;
import android.content.SyncResult;
import android.support.annotation.NonNull;
import android.support.v4.util.ArrayMap;
import android.text.format.DateUtils;

import com.boardgamegeek.R;
import com.boardgamegeek.auth.Authenticator;
import com.boardgamegeek.io.BggService;
import com.boardgamegeek.model.CollectionResponse;
import com.boardgamegeek.model.persister.CollectionPersister;

import java.io.IOException;
import java.util.Date;

import retrofit2.Call;
import retrofit2.Response;
import timber.log.Timber;

/**
 * Syncs the user's collection modified since the date stored in the sync service.
 */
public class SyncCollectionModifiedSince extends SyncTask {
	private SyncResult syncResult;
	private String username;
	private CollectionPersister persister;

	public SyncCollectionModifiedSince(Context context, BggService service) {
		super(context, service);
	}

	@Override
	public int getSyncType() {
		return SyncService.FLAG_SYNC_COLLECTION_DOWNLOAD;
	}

	@Override
	public void execute(@NonNull Account account, @NonNull SyncResult syncResult) {
		this.syncResult = syncResult;
		this.username = account.name;
		AccountManager accountManager = AccountManager.get(context);

		try {
			if (isCancelled()) {
				Timber.i("...cancelled");
				return;
			}

			persister = new CollectionPersister.Builder(context)
				.includeStats()
				.includePrivateInfo()
				.validStatusesOnly()
				.build();

			long date = Authenticator.getLong(accountManager, account, SyncService.TIMESTAMP_COLLECTION_PARTIAL);
			String modifiedSince = BggService.COLLECTION_QUERY_DATE_TIME_FORMAT.format(new Date(date));
			final String formattedDateTime = DateUtils.formatDateTime(context, date, DateUtils.FORMAT_ABBREV_ALL | DateUtils.FORMAT_SHOW_DATE | DateUtils.FORMAT_SHOW_TIME);

			ArrayMap<String, String> options = new ArrayMap<>();
			options.put(BggService.COLLECTION_QUERY_KEY_STATS, "1");
			options.put(BggService.COLLECTION_QUERY_KEY_SHOW_PRIVATE, "1");
			options.put(BggService.COLLECTION_QUERY_KEY_MODIFIED_SINCE, modifiedSince);
			fetchAndPersist(context.getString(R.string.sync_notification_collection_items_since, formattedDateTime), options, "items");

			if (isCancelled()) {
				Timber.i("...cancelled");
				return;
			}
			if (wasSleepInterrupted(2000)) return;

			options.put(BggService.COLLECTION_QUERY_KEY_SUBTYPE, BggService.THING_SUBTYPE_BOARDGAME_ACCESSORY);
			fetchAndPersist(context.getString(R.string.sync_notification_collection_accessories_since, formattedDateTime), options, "accessories");

			Authenticator.putLong(context, SyncService.TIMESTAMP_COLLECTION_PARTIAL, persister.getInitialTimestamp());
		} finally {
			Timber.i("...complete!");
		}
	}

	private void fetchAndPersist(String detail, ArrayMap<String, String> options, final String type) {
		updateProgressNotification(detail);
		Call<CollectionResponse> call = service.collection(username, options);
		try {
			Response<CollectionResponse> response = call.execute();
			if (response.isSuccessful()) {
				CollectionResponse body = response.body();
				if (body != null && body.getItemCount() > 0) {
					int count = persister.save(body.items).getRecordCount();
					syncResult.stats.numUpdates += body.getItemCount();
					Timber.i("...saved %,d records for %,d collection %s", count, body.getItemCount(), type);
				} else {
					Timber.i("...no new collection %s modifications", type);
				}
			} else {
				showError(detail, response.code());
				syncResult.stats.numIoExceptions++;
				cancel();
			}
		} catch (IOException e) {
			showError(detail, e);
			syncResult.stats.numIoExceptions++;
			cancel();
		}
	}

	@Override
	public int getNotificationSummaryMessageId() {
		return R.string.sync_notification_collection_partial;
	}
}

File: app/src/main/java/com/boardgamegeek/service/SyncCollectionModifiedSince.kt
package com.boardgamegeek.service

import android.content.SyncResult
import android.text.format.DateUtils
import com.boardgamegeek.BggApplication
import com.boardgamegeek.R
import com.boardgamegeek.extensions.getDescription
import com.boardgamegeek.extensions.formatDateTime
import com.boardgamegeek.extensions.isCollectionSetToSync
import com.boardgamegeek.io.BggService
import com.boardgamegeek.pref.getCurrentCollectionSyncTimestamp
import com.boardgamegeek.pref.getPartialCollectionSyncLastCompletedAt
import com.boardgamegeek.pref.setPartialCollectionSyncLastCompletedAt
import com.boardgamegeek.repository.CollectionItemRepository
import com.boardgamegeek.util.RemoteConfig
import kotlinx.coroutines.runBlocking
import retrofit2.HttpException
import timber.log.Timber
import java.util.*
import kotlin.time.Duration.Companion.milliseconds

/**
 * Syncs the user's collection modified since the date stored in the sync service.
 */
class SyncCollectionModifiedSince(
    application: BggApplication,
    syncResult: SyncResult,
    private val collectionItemRepository: CollectionItemRepository,
) : SyncTask(application, syncResult) {
    private val fetchPauseMillis = RemoteConfig.getLong(RemoteConfig.KEY_SYNC_COLLECTION_FETCH_PAUSE_MILLIS)

    override val syncType = SyncService.FLAG_SYNC_COLLECTION_DOWNLOAD
    override val notificationSummaryMessageId = R.string.sync_notification_collection_partial

    override fun execute() {
        Timber.i("Starting to sync recently modified collection items")
        try {
            if (isCancelled) {
                Timber.i("Sync task cancelled, aborting")
                return
            }

            if (!prefs.isCollectionSetToSync()) {
                Timber.i("Collection sync not set in preferences, aborting")
                return
            }

            if (syncPrefs.getCurrentCollectionSyncTimestamp() > 0) {
                Timber.i("Currently performing a full sync; skipping incremental sync")
                return
            }

            syncBySubtype()
            if (isCancelled) {
                Timber.i("Sync task cancelled, aborting")
                return
            }

            if (wasSleepInterrupted(fetchPauseMillis.milliseconds)) return

            syncBySubtype(BggService.ThingSubtype.BOARDGAME_ACCESSORY)
            if (isCancelled) {
                Timber.i("Sync task cancelled, aborting")
                return
            }

            syncPrefs.setPartialCollectionSyncLastCompletedAt()
            Timber.i("Syncing recently modified collection completed successfully")
        } catch (e: Exception) {
            Timber.i("Syncing recently modified collection ended with exception:\n$e")
        }
    }

    private fun syncBySubtype(subtype: BggService.ThingSubtype? = null) {
        Timber.i("Starting to sync subtype [${subtype?.code ?: "<none>"}]")
        val lastStatusSync = syncPrefs.getPartialCollectionSyncLastCompletedAt(subtype)
        val lastPartialSync = syncPrefs.getPartialCollectionSyncLastCompletedAt()
        if (lastStatusSync > lastPartialSync) {
            Timber.i("Subtype [${subtype?.code ?: "<none>"}] has been synced in the current sync request; aborting")
            return
        }

        val modifiedSince = BggService.COLLECTION_QUERY_DATE_TIME_FORMAT.format(Date(lastStatusSync))
        val subtypeDescription = subtype.getDescription(context)

        val formattedDateTime = lastStatusSync.formatDateTime(context, flags = DateUtils.FORMAT_ABBREV_ALL or DateUtils.FORMAT_SHOW_DATE or DateUtils.FORMAT_SHOW_TIME)

        updateProgressNotification(context.getString(R.string.sync_notification_collection_since_syncing, subtypeDescription, formattedDateTime))

        val options = mutableMapOf(
            BggService.COLLECTION_QUERY_KEY_STATS to "1",
            BggService.COLLECTION_QUERY_KEY_SHOW_PRIVATE to "1",
            BggService.COLLECTION_QUERY_KEY_MODIFIED_SINCE to modifiedSince,
        )
        subtype?.let { options[BggService.COLLECTION_QUERY_KEY_SUBTYPE] = it.code }

        try {
            val timestamp = System.currentTimeMillis()
            val count = runBlocking { collectionItemRepository.refresh(options, timestamp) }
            syncResult.stats.numUpdates += count.toLong()
            Timber.i("Saved %,d recently modified collection %s", count, subtypeDescription)
            syncPrefs.setPartialCollectionSyncLastCompletedAt(subtype, timestamp)
        } catch (e: Exception) {
            if (e is HttpException) {
                showError(context.getString(R.string.sync_notification_collection_since, subtypeDescription, formattedDateTime), e.code())
            }else{
                showError(context.getString(R.string.sync_notification_collection_since, subtypeDescription, formattedDateTime), e)
            }
            syncResult.stats.numIoExceptions++
            cancel()
        }
    }
}
